File: ml-benchmark/synduce/constraints/sortedlist/failing_slspplus.f
Inductive List = elt Int | cons {Int, List};
Inductive NList = line List | ncons {List, NList};
Inductive CList = sglt List | cat {CList, Int, CList};

c2n =
  let aux = fix (
    \f: NList -> CList -> NList. \pre: NList. \c: CList.
    match c with
      sglt a -> ncons {a, pre}
    | cat {l, _, r} -> f (f pre r) l
    end
  ) in fix (
    \f: CList -> NList. \c: CList.
    match c with
      sglt a -> line a
    | cat {l, _, r} -> aux (f r) l
    end
  );

lsum = fix (
  \f: List -> Int. \xs: List.
  match xs with
    elt x -> x
  | cons {h, t} -> + h (f t)
  end
);

min = \a: Int. \b: Int. if < a b then a else b;
max = \a: Int. \b: Int. if < a b then b else a;

sorted =
  let lmin = fix (
    \f: CList -> Int. \c: CList.
    match c with
      sglt a -> lsum a
    | cat {l, _, r} -> min (f l) (f r)
    end
  ) in let lmax = fix (
    \f: CList -> Int. \c: CList.
    match c with
      sglt a -> lsum a
    | cat {l, _, r} -> max (f l) (f r)
    end
  ) in fix (
  \f: CList -> Bool. \c: CList.
  match c with
    sglt a -> true
  | cat {l, piv, r} -> and (and (< (lmax l) piv) (< piv (lmin r))) (and (f l) (f r))
  end
);

spec = \xs: NList. (fix (
  \f: NList -> {Int, Bool}. \xs: NList.
  match xs with
    line a -> {max 0 (lsum a), >= (lsum a) 0}
  | ncons {h, t} ->
    let res = f t in
      let line_sum = lsum h in
        {
        if and res.2 (>= line_sum 0) then + res.1 line_sum else res.1,
        and res.2 (>= line_sum 0)
        }
  end
) xs).1;

target =
  let list_repr = fix (
    \f: List -> Compress List. \xs: List.
    match xs with
      elt a -> xs
    | cons {h, t} -> cons {h, f t}
    end
  ) in fix (
    \f: CList -> Compress CList. \c: CList.
    match c with
      sglt x -> sglt (list_repr x)
    | cat {l, piv, r} ->
      if <= piv 0 then cat {l, piv, f r} /*Avoid the recursion of l*/
      else cat {f l, piv, f r}
    end
  );

main = \c: CList. if sorted c then spec (c2n (target c)) else 0;

File: ml-benchmark/synduce/constraints/sortedlist/failing_slspplus.ml
type lst =
  | Elt of int
  | Cons of int * lst

type nlist =
  | Line of lst
  | Ncons of lst * nlist

type clist =
  | Sglt of lst
  | Cat of clist * int * clist

let c2n =
  let rec aux pre c =
    match c with
    | Sglt a -> Ncons (a, pre)
    | Cat (l, _, r) -> aux (aux pre r) l
  in
  let rec f c =
    match c with
    | Sglt a -> Line a
    | Cat (l, _, r) -> aux (f r) l
  in
  f

let rec lsum xs =
  match xs with
  | Elt x -> x
  | Cons (h, t) -> h + lsum t

let min a b = if a < b then a else b
let max a b = if a < b then b else a

let sorted =
  let rec lmin c =
    match c with
    | Sglt a -> lsum a
    | Cat (l, _, r) -> min (lmin l) (lmin r)
  in
  let rec lmax c =
    match c with
    | Sglt a -> lsum a
    | Cat (l, _, r) -> max (lmax l) (lmax r)
  in
  let rec f c =
    match c with
    | Sglt _ -> true
    | Cat (l, piv, r) ->
      (lmax l < piv) && (piv < lmin r) && (f l) && (f r)
  in
  f

let spec xs =
  let rec f xs =
    match xs with
    | Line a -> (max 0 (lsum a), lsum a >= 0)
    | Ncons (h, t) ->
      let res = f t in
      let line_sum = lsum h in
      match res with
      | (res1, res2) ->
        let new1 =
          if res2 && line_sum >= 0 then res1 + line_sum else res1
        in
        (new1, res2 && line_sum >= 0)
  in
  let s = f xs in
  match s with
  | (v, _) -> v

val target: clist -> clist compress
let target =
  let rec list_repr xs =
    match xs with
    | Elt _ -> xs
    | Cons (h, t) -> Cons (h, list_repr t)
  in
  let rec f c =
    match c with
    | Sglt x -> Sglt (list_repr x)
    | Cat (l, piv, r) ->
      if piv <= 0 then Cat (l, piv, f r)
      else Cat (f l, piv, f r)
  in
  f

let program c = if sorted c then spec (c2n (target c)) else 0